internal/finalizer/finalizer.go

```diff
@@ -0,0 +1,61 @@
+package finalizer
+
+import (
+	"fmt"
+	corev1 "k8s.io/api/core/v1"
+
+	"context"
+
+	"github.com/matanamar10/namespacelabel-operator/internal/labels"
+
+	"github.com/go-logr/logr"
+	labelsv1 "github.com/matanamar10/namespacelabel-operator/api/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// FinalizerName is a constant used to define the finalizer added to Namespacelabel CRs.
+// This prevents Kubernetes from deleting the CR until the cleanup function completes.
+const finalizerName = "namespacelabels.finalizers.dana.io"
+
+// Ensure ensures that the specified finalizer is added to the Namespacelabel CR if it’s missing.
+// This makes sure that cleanup operations are triggered before deletion.
+func Ensure(ctx context.Context, c client.Client, obj *labelsv1.Namespacelabel, logger logr.Logger) error {
```

Comment:
It would probably be better to generalize this function so that it accepts `obj client.Object` instead of the concrete type of `labelsv1.Namespacelabel`